connectors/src/connectors/microsoft/temporal/activities.ts

```diff
@@ -24,97 +30,217 @@ import type { MicrosoftNodeModel } from "@connectors/lib/models/microsoft";
 import type { WithCreationAttributes } from "@connectors/resources/connector/strategy";
 import { ConnectorResource } from "@connectors/resources/connector_resource";
+import type { MicrosoftNodeType } from "@connectors/resources/microsoft_resource";
 import {
   MicrosoftConfigurationResource,
   MicrosoftNodeResource,
   MicrosoftRootResource,
 } from "@connectors/resources/microsoft_resource";
 import type { DataSourceConfig } from "@connectors/types/data_source_config";
 
-export async function fullSyncActivity({
+export async function getSiteNodesToSync(
+  connectorId: ModelId
+): Promise<MicrosoftNodeResource[]> {
+  const connector = await ConnectorResource.fetchById(connectorId);
+
+  if (!connector) {
+    throw new Error(`Connector with id ${connectorId} not found`);
+  }
+
+  const client = await getClient(connector.connectionId);
+
+  const rootResources =
+    await MicrosoftRootResource.listRootsByConnectorId(connectorId);
+
+  // get root folders and drives and drill down site-root and sites to their
+  // child drives, as MicrosoftNodeData
+  const rootFolderAndDriveNodes: MicrosoftNodeData[] = rootResources.filter(
+    (resource) =>
+      resource.nodeType === "folder" || resource.nodeType === "drive"
+  );
+
+  const rootSiteNodes: MicrosoftNodeData[] = rootResources.filter(
+    (resource) => resource.nodeType === "site"
+  );
+
+  if (rootResources.some((resource) => resource.nodeType === "sites-root")) {
+    const msSites = await getSites(client);
+    rootSiteNodes.push(
+      ...msSites.map(
+        (site): MicrosoftNodeData => ({
+          itemApiPath: getSiteApiPath(site),
+          nodeType: "site",
+        })
+      )
+    );
+  }
+
+  const siteDriveNodes: MicrosoftNodeData[] = (
+    await concurrentExecutor(
+      rootSiteNodes,
+      async (site) => {
+        const msDrives = await getDrives(
+          client,
+          microsoftInternalIdFromNodeData(site)
+        );
+        return msDrives.map((drive) => ({
+          itemApiPath: getDriveApiPath(drive),
+          nodeType: "drive" as const,
+        }));
+      },
+      { concurrency: 5 }
+    )
+  ).flat();
+
+  // remove duplicates
+  const allNodes = [...siteDriveNodes, ...rootFolderAndDriveNodes].reduce(
+    (acc, current) => {
+      const x = acc.find(
+        (item) =>
+          item.nodeType === current.nodeType &&
+          item.itemApiPath === current.itemApiPath
+      );
+      if (!x) {
+        return acc.concat([current]);
+      } else {
+        return acc;
+      }
+    },
+    [] as MicrosoftNodeData[]
+  );
+
+  return MicrosoftNodeResource.batchGetOrCreateFromNodesData(
+    connectorId,
+    allNodes
+  );
+}
+
+export async function markNodeAsVisited(
+  connectorId: ModelId,
+  node: MicrosoftNodeType
+) {
+  const connector = await ConnectorResource.fetchById(connectorId);
+  if (!connector) {
+    throw new Error(`Connector ${connectorId} not found`);
+  }
+
+  await MicrosoftNodeResource.upsert({
```

Comment:
Can we avoid upsert semantics on resources? This should be `updateLastSeenTs`

Can we remove the upsert method entirely I think it's def not in the spirit of Resources as it leaks the underlying model outside of the resource (here it says, the resource is the model and vice-versa which may not be true in the future) cc @flvndvd

Reply:
Understand and agree; will lead to some other refactorings outside this PR but no prob with that
But then here it cannot be only `updateLastSeenTs` since the resource might not exist
Would then be a `getOrCreate` but i'd argue it kind of resembles an `upsert` doesn't it?

Reply:
How would someone know what is used to drive the upsert in a world wher ModelId are not exposed?

Being explicit here has a lot of value as you magically rely on ModelId here which is very hard to grasp from just running the code.